gateway-spi/src/main/java/org/apache/knox/gateway/util/GroupBasedImpersonationProvider.java

```diff
@@ -0,0 +1,257 @@
+        // Map of <hadoop.proxygroup.[VIRTUAL_GROUP].users|groups, group1,group2>
+        Map<String, String> allMatchKeys =
+                getConf().getValByRegex(usersGroupsRegEx);
+
+        for (Map.Entry<String, String> entry : allMatchKeys.entrySet()) {
+            //aclKey = hadoop.proxygroup.[VIRTUAL_GROUP]
+            String aclKey = getAclKey(entry.getKey());
+
+            if (!proxyGroupsAcls.containsKey(aclKey)) {
+                proxyGroupsAcls.put(aclKey, new AccessControlList(
+                        allMatchKeys.get(aclKey + CONF_USERS),
+                        allMatchKeys.get(aclKey + CONF_GROUPS)));
+            }
+        }
+
+        // get hosts per proxygroup
+        allMatchKeys = getConf().getValByRegex(hostsRegEx);
+        for (Map.Entry<String, String> entry : allMatchKeys.entrySet()) {
+            groupProxyHosts.put(entry.getKey(),
+                    new MachineList(entry.getValue()));
+        }
+    }
+
+    private String getAclKey(String key) {
+        int endIndex = key.lastIndexOf('.');
+        if (endIndex != -1) {
+            return key.substring(0, endIndex);
+        }
+        return key;
+    }
+
+    /**
+     * Authorization based on user and group impersonation policies.
+     *
+     * @param user the user information attempting the operation, which includes the real
+     *             user and the effective impersonated user.
+     * @param remoteAddress the remote address from which the user is connecting.
+     * @throws AuthorizationException if the user is not authorized based on the
+     *                                configured impersonation and group policies.
+     */
+    @Override
+    public void authorize(UserGroupInformation user, InetAddress remoteAddress) throws AuthorizationException {
+        authorize(user, remoteAddress, Collections.emptyList());
+    }
+
+    /**
+     * Authorization based on groups that are provided as a function argument
+     *
+     * @param user the user information attempting the operation, which includes the real
+     *             user and the effective impersonated user.
+     * @param groups the list of groups to check for authorization.
+     * @param remoteAddress the remote address from which the user is connecting.
+     * @throws AuthorizationException if the user is not authorized based on the
+     *                                configured impersonation and group policies.
+     */
+    public void authorize(UserGroupInformation user, InetAddress remoteAddress, List<String> groups) throws AuthorizationException {
+
+        /**
+         *  check for proxy user authorization only if PROXYUSER_PREFIX properties exist.
+         *  If proxy is configured then use those properties instead of group-based properties
+         *  given user based configs are more finegrained.
+         */
+        if (doesProxyUserConfigExist) {
+            super.authorize(user, remoteAddress);
+        } else{
```

Comment:
I would expect this to be evaluated in two cases:

1. There is no proxyuser config
2. There is no matching proxyuser config for this specific user

I would not expect this to be an if-else relationship.

Reply:
I mean that the presence of proxyuser config should not preclude the evaluation of the group-level config. If there is user-level config that does not apply to the requesting user (super.authorize() fails), then the group-level config should be evaluated to see if there is matching authorization based on the requesting user's group membership.

Reply:
Probably means catching AuthorizationException from the super.authorize() call, and continuing if there is group-level config. Perhaps, something like
`if(doesProxyUserConfigExist && doesProxyUserConfigExist(user)) {
        try {
            super.authorize(user, remoteAddress);
        } catch (AuthorizationException e) {
            if (!doesProxyGroupConfigExist) {
                throw e;
            }
        }
} 

if (doesProxyGroupConfigExist) {
        checkProxyGroupAuthorization(user, remoteAddress, groups);
}
`

Reply:
Ah, i see thanks for the explanation!